SimplifiedTests/NYPLReaderBookmarksBusinessLogicTests.swift

```diff
@@ -194,8 +198,206 @@ class NYPLReaderBookmarksBusinessLogicTests: XCTestCase {
+    // Make sure server contains no bookmark
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      XCTAssertEqual(bookmarks, nil)
+    }
+    
+    let bookmarkLoc = newBookmarkR2Location(href: "Intro", chapter: "1", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    annotationsMock.failRequest = true
+    guard let _ = bookmarkBusinessLogic.addBookmark(bookmarkLoc) else {
+      XCTFail("Failed to create new bookmark")
+      return
+    }
+    
+    // BookRegistry should have one bookmark even upload failed
+    // While server should not have any bookmarks
+    XCTAssertEqual(self.bookRegistryMock.readiumBookmarks(forIdentifier: self.bookIdentifier).count, 1)
+    
+    annotationsMock.failRequest = false
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      XCTAssertEqual(bookmarks, nil)
+    }
+  }
+
+  // MARK: - Test deleteBookmark/didDeleteBookmark
+  
+  func testDeleteBookmarkWithMatchingBookmark() throws {
+    let firstBookmarkLoc = newBookmarkR2Location(href: "Intro", chapter: "1", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    guard let firstBookmark = bookmarkBusinessLogic.addBookmark(firstBookmarkLoc) else {
+      XCTFail("Failed to create new bookmark")
+      return
+    }
+    
+    let secondBookmarkLoc = newBookmarkR2Location(href: "Chapter 1", chapter: "2", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    guard let _ = bookmarkBusinessLogic.addBookmark(secondBookmarkLoc) else {
+      XCTFail("Failed to create new bookmark")
+      return
+    }
+    
+    // Make sure we have the right amount of bookmarks in BookRegistry
+    XCTAssertEqual(self.bookRegistryMock.readiumBookmarks(forIdentifier: self.bookIdentifier).count, 2)
+    
+    // Deleting one matching bookmark
+    bookmarkBusinessLogic.deleteBookmark(firstBookmark)
+    
+    // BookRegistry should have one bookmark after deletion
+    // Server should not contain the deleted bookmark
+    XCTAssertEqual(self.bookRegistryMock.readiumBookmarks(forIdentifier: self.bookIdentifier).count, 1)
+    annotationsMock.getServerBookmarks(forBook: bookIdentifier,
+                                       publication: nil,
+                                       atURL: nil) { bookmarks in
+      guard let bookmarks = bookmarks else {
+        XCTFail("Failed to get bookmark from server")
+        return
+      }
+      XCTAssertFalse(bookmarks.contains(firstBookmark))
+    }
+  }
+  
+  func testDeleteBookmarkWithMultipleMatchingBookmarks() throws {
+    let firstBookmarkLoc = newBookmarkR2Location(href: "Intro", chapter: "1", progressWithinChapter: 0.1, progressWithinBook: 0.1)
+    guard let firstBookmark = bookmarkBusinessLogic.addBookmark(firstBookmarkLoc),
```

Comment:
hmmm is this a bug in the bookmark logic? should it actually be possible to add the same bookmark 3 times? My expectation would be that the 2nd and 3rd time result in no-ops i.e. one bookmark in total being added (and check at line 308 should be 2). 

If it is a bug, I'd file a ticket for fixing that and change this test accordingly, commenting out for now the parts that are failing until we fix the ticket (also adding a TODO with the ticket).

Reply:
Yeah, that's a bug in the business logic but don't think it would happen in reality since the bookmark button in the reader is turned on/off depends if the bookmark already exist. 
It can be an easy fix by adding a check calling the `bookmarkExisting(location:)` [function](https://github.com/NYPL-Simplified/Simplified-iOS/blob/develop/Simplified/Reader2/BusinessLogic/NYPLReaderBookmarksBusinessLogic.swift#L70) before we add the bookmark to book registry. I can include it in this PR if you don't mind.

Reply:
if it’s that easy, sure, that would be good. That way this test can also be fixed… because right now it seems to suggest a behavior that in theory we don’t want.

Reply:
Fixed